enh: Select platform optimizations at runtime

hershys-aws opened the pull request:
*Description of changes:*
Replace compile-time AWS platform detection with runtime EFA device detection using hwloc. This allows a single binary to work on both AWS and non-AWS environments, automatically enabling optimizations when EFA hardware is present.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

bwbarrett commented on src/nccl_ofi_platform.cpp:
```diff
@@ -0,0 +1,48 @@
+        if (instance) return *instance;
+
+        // Determine if this is an AWS platform.
+        auto is_aws_platform = []() {
```
Why make this a lambda?  We don't have to use every feature.

But also, this isn't a rational way to determine if you're on AWS or not.  `nccl_net_ofi_get_product_name()` will return valid strings on just about any Linux box.  For example, on a VM on my home KVM box, it returns `Standard PC (Q35 + ICH9, 2009)`.

You also can't just match against things that look like instance strings, because the set of instances that can load the plugin are significantly larger than the set that have special entries in the list in platform-aws.c.

hershys-aws replied:
I have change this so that the determination is based on if the provider is EFA instead.

bwbarrett replied:
Let's talk about this 1:1.  You can't use EFA here either.

bwbarrett commented on src/nccl_ofi_platform.cpp:
```diff
@@ -0,0 +1,48 @@
+
+        // NOTE: If we start expanding the number of supported platforms,
+        // move this into a switch statement.
+        if (is_aws_platform()) {
```
ick.  Have the logic for whether or not the aws platform should be loaded *IN* the platform.  Don't have it in two places.

hershys-aws replied:
Re-written.

bwbarrett commented on src/nccl_ofi_platform.cpp:
```diff
@@ -0,0 +1,48 @@
+}
+
+void Platform::device_set_guid(struct fi_info *info, nccl_net_ofi_device_t *device) const {
+        nccl_net_ofi_device_set_guid(info, device);
```
this should not be a global, but part of this class.

bwbarrett reviewed (changes requested):
I stopped after `nccl_ofi_platform.cpp`, because there are some fairly significant architectural approaches to the problem.

The commit message needs a much more detailed description of the change and why you're making it.

hershys-aws commented on tests/unit/aws_platform_mapper.cpp:
```diff
@@ -114,5 +116,29 @@ int main(int argc, char *argv[]) {
+	       test_platform.get_priority());
+
+	// Set parameter using setenv
+	setenv("OFI_NCCL_FORCE_NON_AWS", "true", 1);
```
I have gotten around this by passing a command to the makefile which will run without the AWS platform. Does that work?

hershys-aws commented on src/platform-aws.cpp:
```diff
@@ -1047,3 +1050,33 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
+	hwloc_topology_t topo;
+	if (hwloc_topology_init(&topo) != 0) return false;
+
+	auto topology = std::shared_ptr<hwloc_topology>(topo, hwloc_topology_destroy);
```
Yep. Added commend to clarify as well.

rauteric commented on src/nccl_ofi_net.cpp:
```diff
@@ -1059,6 +1055,7 @@ int nccl_net_ofi_ep_t::release_ep(bool skip_lock, bool force_cleanup)
 	/* Store ref_cnt in local variable in case the endpoint gets deleted */
 	int local_ref_cnt = this->ref_cnt;
 	
+
```
Nit: unrelated whitespace fix

hershys-aws replied:
Fixed.

rauteric commented on src/Makefile.am:
```diff
@@ -48,13 +45,11 @@ else
   # add the tuner sources into the library
-if WANT_PLATFORM_AWS
-    sources +=  \
+  sources +=  \
```
We still want to disable the tuner on non-AWS platforms. Are we confident that works correctly now that the tuner is always being built in?

hershys-aws replied:
I have changed the detection process based on our discussion.

rauteric commented on include/platform-aws.h:
```diff
@@ -72,7 +72,15 @@ class PlatformAWS : public Platform {
+
 private:
+	// Constants
+	static constexpr int BASE_PRIORITY = 100;
```
What is the meaning of `BASE_PRIORITY`?

hershys-aws replied:
I needed a non-zero priority value to assign this class so that it would  be picked when enabled over the default platform. Should this variable be renamed or priority changed?

hershys-aws replied:
Changed to PLATFORM_PRIORITY

bwbarrett replied:
But it's still BASE_PRIORITY.

bwbarrett commented on configure.ac:
```diff
@@ -278,5 +275,5 @@ AC_OUTPUT
 echo "* AWS OFI NCCL plugin has been configured."
 echo "*"
-echo "* Platform-specific optimizations: ${NCCL_OFI_PLATFORM}"
+echo "* Platform optimizations: Runtime detection enabled"
```
I'd list the platforms that were built.

hershys-aws replied:
Done

bwbarrett commented on include/nccl_ofi_param.h:
```diff
@@ -374,4 +374,10 @@ OFI_NCCL_PARAM_UINT(cm_num_rx_buffers, "CM_NUM_RX_BUFFERS", 32);
+ * Force non-AWS platform detection for testing. When set to true,
+ * the AWS platform will report negative priority even on AWS hardware.
+ */
+OFI_NCCL_PARAM(bool, force_non_aws, "FORCE_NON_AWS", false);
```
Why?  Just have `OFI_NCCL_PLATFORM`.  Where `default` is a valid option.

bwbarrett commented on include/nccl_ofi_platform.h:
```diff
@@ -152,7 +152,12 @@ class PlatformManager {
 	 */
-	inline Platform& get_platform() { return *platforms_.rbegin()->second; }
+	inline Platform& get_platform() {
+		Platform& selected = *platforms_.rbegin()->second;
```
This might need a bit of a comment :).

hershys-aws replied:
Oops 😓  Fixed

bwbarrett commented on include/nccl_ofi_platform.h:
```diff
@@ -152,7 +152,12 @@ class PlatformManager {
-	inline Platform& get_platform() { return *platforms_.rbegin()->second; }
+	inline Platform& get_platform() {
+		Platform& selected = *platforms_.rbegin()->second;
+		NCCL_OFI_INFO(NCCL_INIT, "Selected platform: %s with priority %d",
```
`NCCL_NET | NCCL_INIT`.

bwbarrett commented on include/nccl_ofi_platform.h:
```diff
@@ -152,7 +152,12 @@ class PlatformManager {
+	inline Platform& get_platform() {
+		Platform& selected = *platforms_.rbegin()->second;
+		NCCL_OFI_INFO(NCCL_INIT, "Selected platform: %s with priority %d",
+		              selected.get_name(), platforms_.rbegin()->first);
```
for the love of an iterator...

bwbarrett commented on include/platform-aws.h:
```diff
@@ -72,7 +72,15 @@ class PlatformAWS : public Platform {
 	}
 
+	// Determine AWS at runtime
+	bool is_aws() const;
```
Are you actually determining if you're running in AWS, or is this really "system_is_ec2_instance()?

bwbarrett commented on include/platform-aws.h:
```diff
@@ -72,7 +72,15 @@ class PlatformAWS : public Platform {
 		return fields ? fields->func_idx : -EIO;
 	}
 
+	// Determine AWS at runtime
```
This comment is super confusing, since AWS is also the shortened name of this class.

bwbarrett commented on include/platform-aws.h:
```diff
@@ -72,7 +72,15 @@ class PlatformAWS : public Platform {
 private:
+	// Constants
+	static constexpr int BASE_PRIORITY = 100;
+	static constexpr unsigned short AWS_VENDOR_ID = 0x1D0F;
```
This is the Amazon vendor id, not the aws vendor id.  You also should have PCI in there somewhere.

bwbarrett commented on include/platform-aws.h:
```diff
@@ -72,7 +72,15 @@ class PlatformAWS : public Platform {
+	// Constants
+	static constexpr int BASE_PRIORITY = 100;
+	static constexpr unsigned short AWS_VENDOR_ID = 0x1D0F;
+	static constexpr unsigned short EFA_DEV = 0xEFA0;
```
This isn't right.  I'd just have a function `is_efa_device()` that has 0xEFA0 and the 0xFFF mask hard coded, with comments.  Because this method just adds confusion.

bwbarrett commented on src/nccl_ofi_net.cpp:
```diff
@@ -30,9 +30,7 @@
 #include "nccl_ofi_dmabuf.h"
 #include "nccl_ofi_platform.h"
-#ifdef WANT_AWS_PLATFORM
 #include "platform-aws.h"
```
In this world, no code in this file should be aws-specific.

bwbarrett commented on src/nccl_ofi_platform.cpp:
```diff
@@ -21,6 +21,6 @@ void PlatformManager::register_platform(PlatformPtr&& platform) {
 		priority++;
 	}
-
+	NCCL_OFI_INFO(NCCL_INIT, "Adding %s platform with %d priority", name, priority);
```
`NCCL_NET | NCCL_INIT`

bwbarrett commented on src/nccl_ofi_topo.cpp:
```diff
@@ -1093,7 +1093,7 @@ static hwloc_obj_t get_numa_mem_child(hwloc_obj_t node)
  * @return	Pointer to an element of a char array to write device property to.
  *		The array has to be allocated by the caller of this function.
- *		There must be space for at least `MAX_DEV_PROPERTY_LENGTH` 
+ *		There must be space for at least `MAX_DEV_PROPERTY_LENGTH`
```
unrelated change.

bwbarrett commented on src/platform-aws.cpp:
```diff
@@ -1047,3 +1050,38 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
+	}
+
+	hwloc_topology_t topo;
+	if (hwloc_topology_init(&topo) != 0) {
```
why init the topo multiple times per process?

bwbarrett commented on src/platform-aws.cpp:
```diff
@@ -1047,3 +1050,38 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
+	// Use shared_ptr for automatic cleanup of hwloc topology on function exit
+	auto topology = std::shared_ptr<hwloc_topology>(topo, hwloc_topology_destroy);
+	/* HWLOC API changes introduced in version 2.0.0 - same pattern as nccl_ofi_topo.cpp */
+#if (HWLOC_API_VERSION >= 0x00020000)
```
seems like this code is duplicated from topo.c?

hershys-aws replied:
Yeah, AL2 uses an older version of hwloc and so this was needed. I could inline this if desired?

bwbarrett commented on src/platform-aws.cpp:
```diff
@@ -1047,3 +1050,38 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
+	hwloc_obj_t obj = nullptr;
+	while ((obj = hwloc_get_next_pcidev(topo, obj)) != nullptr) {
+		if (obj->attr->pcidev.vendor_id == AWS_VENDOR_ID &&
+		    (obj->attr->pcidev.device_id & 0xFFF0) == EFA_DEV) {
```
I thought we decided that we needed to look for either an ENA or EFA device?  Otherwise, the topo loading for the G* series won't work.

rauteric commented on src/tuner/nccl_ofi_tuner.cpp:
```diff
@@ -63,7 +64,26 @@ static ncclResult_t nccl_ofi_tuner_init(size_t nRanks, size_t nNodes, ncclDebugL
 	 * If both Region and Model based tuner are not supported, log a warning and exit.
 	 */
-	platform_type = nccl_net_ofi_get_product_name();
+	const char* (*get_platform_name)() = (const char* (*)())dlsym(RTLD_DEFAULT, "nccl_ofi_get_selected_platform_name");
```
I'm not sure why you need to use `dlsym` here, instead of just calling the function.

rauteric commented on src/nccl_ofi_topo.cpp:
```diff
@@ -1722,3 +1722,29 @@ struct fi_info *nccl_ofi_topo_next_info_list(nccl_ofi_topo_data_iterator_t *iter
+		return false;
+	}
+
+	auto topology = std::shared_ptr<hwloc_topology>(topo, hwloc_topology_destroy);
```
I didn't understand why you make a shared pointer here, only to never use it.

I think the goal is automatic resource cleanup; would a `unique_ptr` make more sense? And then, for clarity, actually access the topology using the unique ptr, instead of the original raw ptr?

hershys-aws replied:
The shared pointer here was only picked here as it was less verbosity (didn't want to add the decltype and what not. Here would be the equivalent `unique_ptr` code. 
```
std::unique_ptr<hwloc_topology, decltype(&hwloc_topology_destroy)>(...)
```
IMO either works here as the point is just automatic resource cleanup but I can move to unique_ptr if needed.

rauteric commented on src/platform-aws.cpp:
```diff
@@ -1082,3 +1085,17 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
+
+bool PlatformAWS::is_ec2_instance() {
+	const std::string& platform_override = ofi_nccl_platform.get();
+	if (platform_override == "default") {
```
Shouldn't this condition be: platform_override is non-empty and not equal to "aws"? That is, if one forces a different platform, we shouldn't try to enable aws.

AvivBenchorin commented on include/nccl_ofi_param.h:
```diff
@@ -374,4 +374,9 @@ OFI_NCCL_PARAM_UINT(cm_num_rx_buffers, "CM_NUM_RX_BUFFERS", 32);
+/*
+ * Override platform selection. Valid options: "aws", "default", or empty for auto-detection.
+ */
+OFI_NCCL_PARAM(std::string, platform, "PLATFORM", "");
```
Maybe we can use the ENUM param for this instead of a string param, e.g. like the `OFI_NCCL_PROTOCOL` parameter:
```
OFI_NCCL_PARAM_VALUE_SET(PROTOCOL, (SENDRECV)(RDMA))
OFI_NCCL_PARAM(PROTOCOL, protocol, "PROTOCOL", PROTOCOL::SENDRECV);
```
That way you wouldn't have to deal with string comparisons

hershys-aws replied:
Move to this paradigm

AvivBenchorin commented on src/platform-aws.cpp:
```diff
@@ -1082,3 +1083,13 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
 }
+
+bool PlatformAWS::is_ec2_instance() {
+	const std::string& platform_override = ofi_nccl_platform.get();
```
We might want to add some checks that would make sure that the `ofi_nccl_platform` value set in the loaded network library and loaded tuner library are the same so that both libraries are configured correctly. E.g. maybe checking that  `ofi_nccl_platform.get_source() != ParamSource::API` to validate that that `ofi_nccl_platform` wasn't set in the code through a `ofi_nccl_platform.set()` call that could have only happened in either the net library or the tuner library and not the other.

bwbarrett reviewed (commented):
Didn't make it through the full PR, will try to finish today.

hershys-aws commented on configure.ac:
```diff
@@ -278,5 +275,5 @@ AC_OUTPUT
 echo "* AWS OFI NCCL plugin has been configured."
 echo "*"
-echo "* Platform-specific optimizations: ${NCCL_OFI_PLATFORM}"
+echo "* Platform optimizations: AWS, Default"
```
Oh I thought you wanted to list the platforms that were built here: https://github.com/aws/aws-ofi-nccl/pull/995#discussion_r2364444282, but maybe I misunderstood? I can remove it if needed.

bwbarrett replied:
related to comment below, really.  This should just be dynamic.

hershys-aws commented on configure.ac:
```diff
@@ -137,9 +137,6 @@ CHECK_PKG_HWLOC([],
 CHECK_VAR_REDZONE()
 
-NCCL_OFI_PLATFORM="none"
-AS_IF([test "${NCCL_OFI_PLATFORM}" = "none"], [AX_CHECK_PLATFORM_AWS()])
```
The ability to disable AWS platform has been moved to a parameter which would be passed in so I figured this was no longer needed. Is that not the case?

hershys-aws commented on include/nccl_ofi_param.h:
```diff
@@ -374,4 +374,10 @@ OFI_NCCL_PARAM_UINT(cm_num_rx_buffers, "CM_NUM_RX_BUFFERS", 32);
+/*
+ * Override platform selection. Valid options: AWS, DEFAULT, or AUTO for auto-detection.
+ */
+OFI_NCCL_PARAM_VALUE_SET(PLATFORM, (AUTO)(AWS)(DEFAULT))
```
As part of @AvivBenchorin feedback I moved to an enum and so auto does the automatic detection, and default is the platform when we disable AWS. By doing this we can force an AWS platform or default platform or just set it to auto detect (default value).

hershys-aws replied:
Moved this to back to taking in a string instead.

hershys-aws commented on include/platform-aws.h:
```diff
@@ -23,7 +23,7 @@
 public:
 	const char* get_name() const override { return "AWS"; }
-	int get_priority() override { return 100; }
+	int get_priority() override { return PLATFORM_PRIORITY * (is_ec2_instance() ? 1 : -1); }
```
The std::map orders the platform based on priority. As Default will always be 0 we just need the priority to be negative if we aren't on an aws ec2 instance so that we disable the aws platform. This is evaluated when we create thee platform manager object and so is automatically calculated and set.

hershys-aws commented on include/nccl_ofi_platform.h:
```diff
@@ -167,9 +164,21 @@ class PlatformManager {
+	 *
+	 * @param	platform	Platform instance to register (moved)
+	 */
+	void register_platform(PlatformPtr&& platform);
```
The only reason there is a rvalue operator there is so that it can be moved to be owned by the platform manager explicitly. This can be changed if you see a need for the platform to "run functions" extensively. The pointer part can be removed if that helps, but most of this "just works" in the background in the sense that each platform is added to the platform manager and has a priority set and finally the highest priority gets picked. Adding future platforms is as simple as just doing `register_platform` with the platform as an rvalue. We can discuss what part to simplify if needed. The only thing which probably needs work (when have multiple platforms) is how to deal with differing platforms having the same priority.

hershys-aws replied:
Based on our chat, I have moved this to only hold a single platform and be calculated once. Hopefully this implementation is cleaner/simpler.

bwbarrett commented on src/platform-aws.cpp:
```diff
@@ -1082,3 +1083,18 @@ void PlatformAWS::sort_rails(struct fi_info **info_list, size_t num_rails, size_
+
+	PLATFORM platform_override = ofi_nccl_platform.get();
+	if (platform_override != PLATFORM::AUTO) {
+		return platform_override == PLATFORM::AWS;
```
Why should this be in each platform file?  What's the point of the platform manager code if you're just going to have a bunch of the selection logic elsewhere?

bwbarrett commented on tests/unit/aws_platform_mapper.cpp:
```diff
@@ -115,5 +116,13 @@ int main(int argc, char *argv[]) {
 
+	// Test current environment setting
+	TestablePlatformAWS test_platform;
+	printf("  Current environment - is_ec2_instance(): %s, priority: %d\n",
```
Not really sure what this check does in terms of a unit test?

hershys-aws replied:
The original point of this was to exercise this path, and those function with the idea that it could be tested on the compute nodes for "verification" that the logic works. I will remove it as it doesn't seem that this is the right place to put this upon further thought...

bwbarrett replied:
We should definitely exercise this path, but you need some way of doing that which is clearly pass / fail.  I think the only rational way to do that is to have a test that will skip (return 77) normally, but use an environment variable set by our test harness to say "yes, this is an ec2 instance"

bwbarrett commented on tests/unit/platform_manager.cpp:
```diff
@@ -10,22 +10,24 @@
 public:
 	const char* get_name() const override { return "TestPlatform"; }
-	int get_priority() override { return 10; }
+	int get_priority() override { return 200; }
```
This does not seem like a good design pattern.

hershys-aws replied:
I had to keep this as part of the priority value approach...not sure how else to modify/change this?

hershys-aws commented on src/nccl_ofi_topo.cpp:
```diff
@@ -1722,3 +1722,29 @@ struct fi_info *nccl_ofi_topo_next_info_list(nccl_ofi_topo_data_iterator_t *iter
+bool nccl_ofi_topo_has_efa_ena_devices()
+{
+	hwloc_topology_t topo;
+	if (hwloc_topology_init(&topo) != 0) {
```
I have moved away to adding a singleton class (reasoning as the topology should only be initialized once) and having that take care of the lifecycle of the topology object. Eventual goal is to pull other functions into that implementation (probably good as we should have this as part of C++ refactor anyways).

bwbarrett commented on configure.ac:
```diff
@@ -278,5 +278,7 @@ AC_OUTPUT
-echo "* Platform-specific optimizations: ${NCCL_OFI_PLATFORM}"
+AS_IF([test "${NCCL_OFI_PLATFORM}" = "none"],
+      [echo "* Platform Optimizations: DISABLED"],
+      [echo "* Platform Optimizations: ${NCCL_OFI_PLATFORM}"])
```
This isn't quite what I asked for.  "DISABLED" is really wrong, since default will still be built.  Just build up a list of the enabled platforms (including the default plugin, of course).

bwbarrett commented on include/nccl_ofi_param.h:
```diff
@@ -374,4 +374,9 @@ OFI_NCCL_PARAM_UINT(cm_num_rx_buffers, "CM_NUM_RX_BUFFERS", 32);
+/*
+ * Override platform selection. Valid options: "AWS", "Default", or empty string for auto-detection.
+ */
+OFI_NCCL_PARAM(std::string, platform_override, "PLATFORM_OVERRIDE", "");
```
Why does "override" make sense?  It doesn't match all the other option we have (like OFI_NCCL_PROGRESS_MODEL above).  This should just be OFI_NCCL_PLATFORM.

bwbarrett commented on include/nccl_ofi_topo.h:
```diff
@@ -310,4 +312,49 @@ struct fi_info *nccl_ofi_topo_next_info_list(nccl_ofi_topo_data_iterator_t *iter
  */
 int nccl_ofi_topo_write_nccl_topology(nccl_ofi_topo_t *topo, FILE *file);
 
+class TopologyManager {
```
I strongly object to half-refactoring this code.  There's also no need for a topology manager object.

hershys-aws committed:
topology: split logic from nccl_ofi_topo_create

Refactored nccl_ofi_topo_create so that the filtering is done at a later
step. This is in preparation of dynamically selecting platform as we
need the topology to determine the platform for aws optimizations. Added
function to check any EFA or ENA devices exist in the topology for
platform detection.

Signed-off-by: Hershel Shah <hershys@amazon.com>

hershys-aws committed:
refactor: Simplify PlatformManager logic

Change PlatformManager logic to work off a compare and set instead of
using the std::map. AWS specific logic moved from net to platform file
and parameter to override parameter for runtime-detection has been
introduced. Refactored unit tests accordingly.

Signed-off-by: Hershel Shah <hershys@amazon.com>

hershys-aws committed:
usability: Runtime selection of platform optimizations

Added logic to dynamically select between Default and PlatformAWS. Move
creation of topology to the net file and refactored rdma plugin to take
in the topology. Added topo parameter to PlatformAWS constructor to
determine if efa/ena devices are present in the topology for priority
calculation.

Signed-off-by: Hershel Shah <hershys@amazon.com>

bwbarrett reviewed (commented):
Please make sure your commit messages conform to the standard.  Some of your title lines are well over the 56 character limit.

hershys-aws commented on src/tuner/nccl_ofi_tuner.cpp:
```diff
@@ -59,11 +60,22 @@ static ncclResult_t nccl_ofi_tuner_init(size_t nRanks, size_t nNodes, ncclDebugL
+
+	/* Default platform or other non-AWS platforms should use internal tuner */
 	if (platform_type == NULL) {
 		NCCL_OFI_WARN("NCCL_OFI_TUNER is not available because platform type is unavailable.");
```
Changed as I have to make fixes regardless 😄
